### DISCUSSION TITLE
Move execution of trailingSlash logic after _middleware execution

### DISCUSSION THREAD

**Discussion by andriiradkevych** [freshframework/fresh#2338] [Ideas]
**_middleware.ts**
```javascript
import { FreshContext } from '$fresh/server.ts';

export function handler(req: Request, ctx: FreshContext) {

  if (
    (ctx.destination === 'route' || ctx.destination === 'notFound') &&
    /[A-Z]/.test(ctx.url.pathname)
  ) {
    ctx.url.pathname = ctx.url.pathname.toLowerCase();

    return Response.redirect(ctx.url, 301);
  }

  return ctx.next();
}

```

**fresh.config.ts**
```javascript
import { defineConfig } from '$fresh/server.ts';

export default defineConfig({
  port: 3000,
  router: {
    trailingSlash: true,
  },
});
```

For navigation to url `/Page` , I've got next redirect path:

308 permanent redirect - your redirection for `trailingSlash: true` => `/page`
301 - my redirect logic on middleware => `/page/`

it shouldn't be like this , I should have in `/Page` only 1 redirect => `/page/` , and this logic transformation I can do myself , 
or there should be next logic . On middleware I can transform url :

```javascript
import { FreshContext } from '$fresh/server.ts';

export function handler(req: Request, ctx: FreshContext) {
  if (
    (ctx.destination === 'route' || ctx.destination === 'notFound') &&
    /[A-Z]/.test(ctx.url.pathname)
  ) {
    ctx.url.pathname = ctx.url.pathname.toLowerCase();
  }

  return ctx.next();
}
```

without any redirection , and then after middleware you execute trailing slash redirect logic , and I will get only 308 permanent redirect `/page/` 

or maybe it is possible to make it , like 

```javascript
import { defineConfig } from '$fresh/server.ts';

export default defineConfig({
  port: 3000,
  router: {
    changeUrlCallback: (url) => {
         if (/[A-Z]/.test(ctx.url.pathname)){
           url.pathname = ctx.url.pathname.toLowerCase();
         }
    }
  },
});
```

and you check is url changed by callback , and if so , then do redirect 

or just add additional properties pls :

```
  router: {
    trailingSlash: true,
    transformUrlToLowerCase: true,
    redirectFromHttpToHttps: true
  },
```

**Comment by nnmrts** [comment]
I'm not sure what the question here is. Why are two redirects a problem?

Anyways, two options to avoid an additional redirect:

- append trailing slash in your middleware
	- `ctx.url.pathname.toLowerCase().replace(/\/?$/, "/")`
- add a [`routeOverride` url pattern](https://fresh.deno.dev/docs/concepts/routing) (probably overkill)
	- `"/:page(Page/page)"`
